### ISSUE TITLE
Explanation of Tensorboard Validation Graphs

### ISSUE THREAD

**Issue by mittski** [mlfoundations/open_clip#733]
Would anyone more knowledgeable be willing to share some insight as to what all the provided validation graphs mean?

I understand the premise of Validation, and I'm pretty sure first graph Clip_val_loss reflects that as is.. but then there's image_to_text_R@1 through 5, their mean rank, and text-to image ranks1-5 and their mean as well. Any clarification would be nice, for a chance to better reflect on this information given!

**Comment by gabrielilharco** [comment]
Hey @mittski. These are image-to-text and text-to-image retrieval metrics. R means recall. For example, image_to_text_R@5 means the percentage of samples for which the matching text was accurately retrieved on the top-5 options according to the model. Rank refers to the rank of the correctly paired sample according to the model, among all choices. Mean and median rank are just aggregates of that across all samples. Let me know if that helps or if you have any other questions!

**Comment by rom1504** [comment]
You may find it useful to read the metrics code
https://github.com/LAION-AI/CLIP_benchmark/tree/main/clip_benchmark/metrics

**Comment by gabrielilharco** [comment]
See also https://github.com/mlfoundations/open_clip/blob/main/src/training/train.py#L360

**Comment by mittski** [comment]
Thanks for the clarification @gabrielilharco , and links, that does demystify a lot

> ... image_to_text_R@5 means the percentage of images for which the matching text was accurately retrieved on the top-5 options according to the model. Rank refers to the rank of the correctly paired sample according to the model, among all choices. 

Looking at the code, it's not entirely clear how a 'match' is determined: within each batch all the text/image embeddings are compared, and a 'hit' is classified as..? Comparing all embeddings, if closest the one is it's true pairing: its a hit?

If that's the case.. I wonder how informative it is, if the magnitude of the difference isn't considered, but only the relative difference between neighbors: that you could finetune on a dataset and get good rank on ID data, due to trivially small but accurate differences in their vectors, but it still be undertrained relative to the rest of the model which have differences encoded at larger or smaller orders of magnitude?

**Comment by gabrielilharco** [comment]
Hi @mittski. Re. what a match is, maybe a simple example would help. Let's say we have a dataset with 5 image-text pairs, ((i1, t1), (i2, t2), (i3, t3), (i4, t4), (i5, t5)). When the model does image-to-text retrieval for the first image, i1, it computes the distance to all text features and sort them based on that. Let's say that according to the model the order is (t3, t2, t1, t5, t4). In this case, the rank would be 3 (since t1 is in the 3rd position). For that sample, R@1 would be zero, R@5 would be one. These metrics would be aggregated for all images.  

I'm not sure if I fully understand your last point, but indeed these are discrete metrics (much like accuracy), and depend only on the rank of the samples (not on the magnitude of score differences). If you want a more 'soft' measure, you can always look at validation accuracy

**Comment by mittski** [comment]
thanks for clarification, that makes a lot more sense
Yeah, I think that best consideration of these metrics would be in context with the val accuracy. My concern may have been cart-before-the-horse, I don't know if in practice the case of "High val accuracy, low rank ratings" / "low val accuracy, high rank ratings" may not ever occur - and if it does, you have the val accuracy graph to distinguish between the two

**Comment by gabrielilharco** [comment]
You would need to modify https://github.com/mlfoundations/open_clip/blob/main/src/training/train.py#L360 to take this into account. Right now the code assumes there's a 1-1 pairing between images and text such (the i'th image is paired with the i'th text). If you have another kind of relationship, you'd need to modify the code accordingly. For R@k, instead of checking whether the ranking of the best prediction is < k, you'd can check whether any of the top-k predictions are in the set of matches
